File: ports/php/src/Security/Covert/TagBlockPayload.php
<?php

declare(strict_types=1);

namespace UnicodePhp\Security\Covert;

use UnicodePhp\Security\ClassificationKind;

/// Sub-threat variant for the tag-block-payload detector. Each variant carries
/// its attribution payload and reports a fixture-row tag string.
interface TagBlockSubThreat
{
    public function tag(): string;
}

/// Input is pure tags and the decoder produces ≥ 1 printable byte.
final class TagBlockDirectAscii implements TagBlockSubThreat
{
    public function __construct(public readonly string $decoded)
    {
    }

    public function tag(): string
    {
        return 'DirectAscii';
    }
}

/// E0001 followed by ≥ 1 further tag char.
final class TagBlockLanguageTagRevival implements TagBlockSubThreat
{
    public function __construct(
        public readonly int $langTagPos,
        public readonly string $decodedTail,
    ) {
    }

    public function tag(): string
    {
        return 'LanguageTagRevival';
    }
}

/// Tag chars interleaved with non-tag codepoints.
final class TagBlockMixedBlock implements TagBlockSubThreat
{
    public function __construct(
        public readonly int $tagCount,
        public readonly int $totalCps,
    ) {
    }

    public function tag(): string
    {
        return 'MixedBlock';
    }
}

/// Fallback for isolated single tag-block codepoints.
final class TagBlockBareTagPresent implements TagBlockSubThreat
{
    public function __construct(public readonly int $tagCp)
    {
    }

    public function tag(): string
    {
        return 'BareTagPresent';
    }
}

/// Structured verdict over an input codepoint sequence.
final class TagBlockVerdict
{
    /**
     * @param list<int> $tagPositions
     */
    public function __construct(
        public readonly ClassificationKind $kind,
        public readonly ?TagBlockSubThreat $sub,
        public readonly array $tagPositions,
        public readonly string $recoveredAscii,
    ) {
    }
}

/// Detection of invisible payloads encoded in the Unicode tag block
/// U+E0000..U+E007F under the decoder `tag(c) = c + 0xE0000` for c in
/// [0x20, 0x7E]. Every occurrence is reportable; the detector attributes the
/// kind of use.
final class TagBlockPayload
{
    /// True iff `cp` is in the Unicode tag block U+E0000..U+E007F.
    public static function isTagCharacter(int $cp): bool
    {
        return $cp >= 0xE0000 && $cp <= 0xE007F;
    }

    public static function isLanguageTag(int $cp): bool
    {
        return $cp === 0xE0001;
    }

    public static function isCancelTag(int $cp): bool
    {
        return $cp === 0xE007F;
    }

    /// Decode a tag-block codepoint to its ASCII correspondent, or `null` for
    /// tag codepoints outside the printable-ASCII range and non-tag codepoints.
    public static function tagToAscii(int $cp): ?string
    {
        if ($cp >= 0xE0020 && $cp <= 0xE007E) {
            return chr($cp - 0xE0000);
        }
        return null;
    }

    /**
     * @param list<int> $input
     * @param list<int> $positions
     */
    private static function decodeTagRun(array $input, array $positions): string
    {
        $s = '';
        $len = count($input);
        foreach ($positions as $p) {
            if ($p < $len) {
                $c = self::tagToAscii($input[$p]);
                if ($c !== null) {
                    $s .= $c;
                }
            }
        }
        return $s;
    }

    /**
     * @param list<int> $input
     * @param list<int> $tagPositions
     */
    private static function hasLanguageTagPrefix(array $input, array $tagPositions): ?int
    {
        if ($tagPositions === []) {
            return null;
        }
        $langPos = $tagPositions[0];
        if ($langPos >= count($input)) {
            return null;
        }
        if (self::isLanguageTag($input[$langPos]) && count($tagPositions) >= 2) {
            return $langPos;
        }
        return null;
    }

    /**
     * @param list<int> $input
     * @param list<int> $tagPositions
     */
    private static function pickSubThreat(array $input, array $tagPositions, string $decoded): TagBlockSubThreat
    {
        $langPos = self::hasLanguageTagPrefix($input, $tagPositions);
        if ($langPos !== null) {
            $tail = array_values(array_filter($tagPositions, static fn (int $p): bool => $p !== $langPos));
            return new TagBlockLanguageTagRevival($langPos, self::decodeTagRun($input, $tail));
        }
        $allTags = true;
        foreach ($input as $cp) {
            if (!self::isTagCharacter($cp)) {
                $allTags = false;
                break;
            }
        }
        if ($allTags && $decoded !== '') {
            return new TagBlockDirectAscii($decoded);
        }
        if (count($input) > count($tagPositions)) {
            return new TagBlockMixedBlock(count($tagPositions), count($input));
        }
        return new TagBlockBareTagPresent($input[$tagPositions[0]]);
    }

    /**
     * @param list<int> $input
     */
    public static function detect(array $input): TagBlockVerdict
    {
        $tagPositions = [];
        foreach ($input as $i => $cp) {
            if (self::isTagCharacter($cp)) {
                $tagPositions[] = $i;
            }
        }

        if ($tagPositions === []) {
            return new TagBlockVerdict(ClassificationKind::Clear, null, [], '');
        }

        $decoded = self::decodeTagRun($input, $tagPositions);
        $sub = self::pickSubThreat($input, $tagPositions, $decoded);

        return new TagBlockVerdict(ClassificationKind::Hazard, $sub, $tagPositions, $decoded);
    }
}

File: ports/lua/src/unicode_lua/security/covert/tag_block_payload.lua
-- Detection of invisible payloads encoded in the Unicode tag block
-- U+E0000..U+E007F.  Every occurrence is reportable; the detector attributes
-- the kind of use (direct payload, language-tag prefix, mixed-in text, or
-- isolated single tag).  Positions are 0-based codepoint offsets.

local calculus = require("unicode_lua.security.calculus")
local ClassificationKind = calculus.ClassificationKind

local M = {}

function M.is_tag_character(cp)
  return cp >= 0xE0000 and cp <= 0xE007F
end

function M.is_language_tag(cp)
  return cp == 0xE0001
end

function M.is_cancel_tag(cp)
  return cp == 0xE007F
end

-- Decode a tag-block codepoint to its ASCII correspondent, or nil.
function M.tag_to_ascii(cp)
  if cp >= 0xE0020 and cp <= 0xE007E then
    return string.char(cp - 0xE0000)
  end
  return nil
end

-- `positions` is a 0-based list.
local function decode_tag_run(input, positions)
  local s = {}
  for _, p in ipairs(positions) do
    if p < #input then
      local c = M.tag_to_ascii(input[p + 1])
      if c ~= nil then
        s[#s + 1] = c
      end
    end
  end
  return table.concat(s)
end

local function all_tag_characters(input)
  for _, cp in ipairs(input) do
    if not M.is_tag_character(cp) then
      return false
    end
  end
  return true
end

local function has_language_tag_prefix(input, tag_positions)
  if #tag_positions == 0 then
    return nil
  end
  local lang_pos = tag_positions[1]
  if lang_pos >= #input then
    return nil
  end
  if M.is_language_tag(input[lang_pos + 1]) and #tag_positions >= 2 then
    return lang_pos
  end
  return nil
end

local function pick_sub_threat(input, tag_positions, decoded)
  local lang_pos = has_language_tag_prefix(input, tag_positions)
  if lang_pos ~= nil then
    local tail = {}
    for _, p in ipairs(tag_positions) do
      if p ~= lang_pos then
        tail[#tail + 1] = p
      end
    end
    return { tag = "LanguageTagRevival", lang_tag_pos = lang_pos, decoded_tail = decode_tag_run(input, tail) }
  end
  if all_tag_characters(input) and decoded ~= "" then
    return { tag = "DirectAscii", decoded = decoded }
  end
  if #input > #tag_positions then
    return { tag = "MixedBlock", tag_count = #tag_positions, total_cps = #input }
  end
  return { tag = "BareTagPresent", tag_cp = input[tag_positions[1] + 1] }
end

-- Returns { kind, sub, tag_positions (0-based), recovered_ascii }.
function M.detect(input)
  local tag_positions = {}
  for i = 1, #input do
    if M.is_tag_character(input[i]) then
      tag_positions[#tag_positions + 1] = i - 1
    end
  end

  if #tag_positions == 0 then
    return { kind = ClassificationKind.Clear, sub = nil, tag_positions = {}, recovered_ascii = "" }
  end

  local decoded = decode_tag_run(input, tag_positions)
  local sub = pick_sub_threat(input, tag_positions, decoded)

  return { kind = ClassificationKind.Hazard, sub = sub, tag_positions = tag_positions, recovered_ascii = decoded }
end

return M
